autorest/azure/example/main.go

```diff
@@ -104,27 +150,92 @@ func getResourceGroups(client *autorest.Client) (*string, error) {
 		return nil, err
 	}
 
+	value := struct {
+		ResourceGroups []struct {
+			Name string `json:"name"`
+		} `json:"value"`
+	}{}
+
 	defer resp.Body.Close()
-	contents, err := ioutil.ReadAll(resp.Body)
+	dec := json.NewDecoder(resp.Body)
+	err = dec.Decode(&value)
 	if err != nil {
 		return nil, err
 	}
 
-	contentsString := string(contents)
+	var names = make([]string, len(value.ResourceGroups))
+	for i, name := range value.ResourceGroups {
+		names[i] = name.Name
+	}
+	return names, nil
+}
 
-	return &contentsString, nil
+func saveToken(spt azure.Token) {
```

Comment:
I think a refreshfunc should return an error. So I think you should define

    type TokenRefreshCallback func (azure.Token) error

and use that in the `SetRefreshCallback()`. I think the error returned from the callback (i.e. failure to save the token to a file for instance) can be handled by the `Refresh()` function in the autorest ––simply because when such a case happens, autorest is the caller and it is the package being executed.

Otherwise your only option in this callback is to a) panic b) log.Fatal/os.Exit(1). Neither are good nor idiomatic.